src/associations/has-many.ts

```diff
@@ -0,0 +1,834 @@
+
+    // Get singular and plural names
+    // try to uppercase the first letter, unless the model forbids it
+    const plural = upperFirst(this.options.name.plural);
+    const singular = upperFirst(this.options.name.singular);
+
+    this.accessors = {
+      get: `get${plural}`,
+      set: `set${plural}`,
+      addMultiple: `add${plural}`,
+      add: `add${singular}`,
+      create: `create${singular}`,
+      remove: `remove${singular}`,
+      removeMultiple: `remove${plural}`,
+      hasSingle: `has${singular}`,
+      hasAll: `has${plural}`,
+      count: `count${plural}`,
+    };
+
+    this.#mixin(source.prototype);
+  }
+
+  static associate<
+    S extends Model,
+    T extends Model,
+    SourceKey extends AttributeNames<S>,
+    TargetKey extends AttributeNames<T>,
+    >(
+    secret: symbol,
+    source: ModelStatic<S>,
+    target: ModelStatic<T>,
+    options: HasManyOptions<SourceKey, TargetKey> = {},
+    parent?: Association<any>,
+  ): HasMany<S, T, SourceKey, TargetKey> {
+
+    return defineAssociation<
+      HasMany<S, T, SourceKey, TargetKey>,
+      HasManyOptions<SourceKey, TargetKey>,
+      NormalizedHasManyOptions<SourceKey, TargetKey>
+    >(HasMany, source, target, options, parent, normalizeBaseAssociationOptions, normalizedOptions => {
+      // self-associations must always set their 'as' parameter
+      if (isSameInitialModel(source, target)
+        // use 'options' because this will always be set in 'newOptions'
+        && (!options.as || !options.inverse?.as || options.as === options.inverse.as)) {
+        throw new AssociationError('Both options "as" and "inverse.as" must be defined for hasMany self-associations, and their value must be different.');
+      }
+
+      return new HasMany(secret, source, target, normalizedOptions, parent);
+    });
+  }
+
+  #mixin(mixinTargetPrototype: Model) {
+    mixinMethods(
+      this,
+      mixinTargetPrototype,
+      ['get', 'count', 'hasSingle', 'hasAll', 'set', 'add', 'addMultiple', 'remove', 'removeMultiple', 'create'],
+      {
+        hasSingle: 'has',
+        hasAll: 'has',
+        addMultiple: 'add',
+        removeMultiple: 'remove',
+      },
+    );
+  }
+
+  /**
+   * Get everything currently associated with this, using an optional where clause.
+   *
+   * @param instances source instances
+   * @param options find options
+   */
+  async get(instances: S, options?: HasManyGetAssociationsMixinOptions<T>): Promise<T[]>;
+  async get(instances: S[], options?: HasManyGetAssociationsMixinOptions<T>): Promise<Map<any, T[]>>;
+  async get(instances: S | S[], options: HasManyGetAssociationsMixinOptions<T> = {}): Promise<T[] | Map<any, T[]>> {
+    let isManyMode = true;
+    if (!Array.isArray(instances)) {
+      isManyMode = false;
+      instances = [instances];
+    }
+
+    const findOptions: FindOptions = { ...options };
+
+    const where = Object.create(null);
+
+    // FIXME: scopes should be combined using AND instance of overwriting.
```

Comment:
```suggestion
    // TODO: scopes should be combined using AND instance of overwriting.
```
Is this for this PR, before the release of v7 or after?